fitbenchmarking/parsing/horace_parser.py

```diff
@@ -0,0 +1,194 @@
+"""
+This file implements a parser for Horace problem sets.
+"""
+import os
+import typing
+
+import matlab
+import numpy as np
+
+from fitbenchmarking.parsing.fitbenchmark_parser import FitbenchmarkParser
+from fitbenchmarking.utils.exceptions import ParsingError
+from fitbenchmarking.utils.matlab_engine import ENG as eng
+from fitbenchmarking.utils.matlab_engine import add_persistent_matlab_var
+
+
+def horace_on():
+    """
+    Turning Horace and SpinW on in matlab
+    """
+    if "HORACE_LOCATION" in os.environ and "SPINW_LOCATION" in os.environ:
+        horace_location = os.environ["HORACE_LOCATION"]
+        spinw_location = os.environ["SPINW_LOCATION"]
+        eng.evalc("restoredefaultpath")
+        eng.evalc(f"addpath('{horace_location}')")
+        eng.evalc("horace_on")
+        eng.evalc(f"addpath('{spinw_location}')")
+        eng.evalc("spinw_on")
+    elif "HORACE_LOCATION" not in os.environ and \
+         "SPINW_LOCATION" in os.environ:
+        raise ParsingError('Could not parse SpinW problem. Please ensure '
+                           'that HORACE_LOCATION is specfied as a environment '
+                           'variable')
+    elif "HORACE_LOCATION" in os.environ and \
+         "SPINW_LOCATION" not in os.environ:
+
+        raise ParsingError('Could not parse SpinW problem. Please ensure '
+                           'that SPINW_LOCATION is specfied as a environment '
+                           'variable')
+    else:
+        eng.evalc("horace_on")
+
+
+horace_on()
+
+
+class HoraceParser(FitbenchmarkParser):
+    """
+    Parser for a Horace problem definition file.
+    """
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+        # A container for improving efficiency in function call if calling
+        # with full input data
+        self._horace_x: typing.Optional[np.ndarray] = None
+        self._horace_w: typing.Optional[str] = None
+        self._horace_msk: typing.Optional[str] = None
+        self._horace_path: typing.Optional[str] = None
+
+    def _get_data_points(self, data_file_path):
+        """
+        Get the data points of the problem from the data file.
+
+        :param data_file_path: The path to the file to load the points from
+        :type data_file_path: str
+
+        :return: data
+        :rtype: dict<str, np.ndarray>
+        """
+        wye_data_f = self._parse_function(self._entries['wye_function'])
+        path = os.path.join(os.path.dirname(self._filename),
+                            wye_data_f[0]['matlab_script'])
+        eng.addpath(os.path.dirname(path))
+        func_name = os.path.basename(path).split('.', 1)[0]
```

Comment:
See [pathlib](https://docs.python.org/3/library/pathlib.html?highlight=pathlib#module-pathlib) for a more modern/less verbose way of doing this.